File: openclaw_republic/bus/event_log.py
"""结构化事件日志记录器。

所有 Agent action 统一记录为结构化事件，
含 emotion, intensity 等字段，
直接对标 PRD §4 的 WebSocket 事件格式。
"""

from __future__ import annotations

from datetime import datetime

from openclaw_republic.schemas.events import BaseEvent, EventAction


class EventLogger:
    """结构化事件日志记录器。

    记录所有 Agent 动作事件，支持按条件查询和
    WebSocket 推送格式导出。
    """

    def __init__(self) -> None:
        self._events: list[BaseEvent] = []

    def log(self, event: BaseEvent) -> None:
        """记录一条事件。

        Args:
            event: 要记录的事件。
        """
        self._events.append(event)

    def get_events(
        self,
        source_agent: str | None = None,
        action: EventAction | None = None,
        since: datetime | None = None,
    ) -> list[BaseEvent]:
        """按条件查询事件。

        所有参数均为可选过滤条件，多个条件为 AND 逻辑。

        Args:
            source_agent: 按发出事件的 Agent 角色名过滤。
            action: 按事件动作类型过滤。
            since: 只返回此时间之后的事件。

        Returns:
            符合条件的事件列表。
        """
        results: list[BaseEvent] = []
        for event in self._events:
            if source_agent is not None and event.source_agent != source_agent:
                continue
            if action is not None and event.action != action:
                continue
            if since is not None:
                event_ts = event.timestamp.replace(tzinfo=None)
                since_ts = since.replace(tzinfo=None)
                if event_ts < since_ts:
                    continue
            results.append(event)
        return results

    def export_for_websocket(self) -> list[dict[str, object]]:
        """导出为 WebSocket 推送格式。

        Returns:
            事件字典列表，每个字典包含事件的所有字段。
        """
        return [event.model_dump(mode="json") for event in self._events]

    @property
    def count(self) -> int:
        """已记录的事件总数。"""
        return len(self._events)

    def clear(self) -> None:
        """清空所有事件记录。"""
        self._events.clear()

File: backend/src/bus/event-log.ts
/**
 * 结构化事件日志记录器。
 *
 * 所有 Agent action 统一记录为结构化事件，
 * 直接对标 PRD 的 WebSocket 事件格式。
 */

import { BaseEvent, EventAction } from '../schemas/events';

const MAX_EVENT_LOG_SIZE = 10_000;

export class EventLogger {
  private _events: BaseEvent[];

  constructor() {
    this._events = [];
  }

  public log(event: BaseEvent): void {
    this._events.push(event);
    // Bug 35+54 fix: 防止无限增长，超过容量则原地通过 splice 丢弃旧数据，保持数组引用
    if (this._events.length > MAX_EVENT_LOG_SIZE) {
      this._events.splice(0, MAX_EVENT_LOG_SIZE / 2);
    }
  }

  public get_events(filters?: {
    source_agent?: string;
    action?: EventAction;
    since?: Date;
  }): BaseEvent[] {
    return this._events.filter(event => {
      if (filters?.source_agent && event.source_agent !== filters.source_agent) {
        return false;
      }
      if (filters?.action && event.action !== filters.action) {
        return false;
      }
      if (filters?.since) {
        // Bug 48 fix: handle both Date objects and ISO string timestamps
        const eventTime = event.timestamp instanceof Date
          ? event.timestamp.getTime()
          : new Date(event.timestamp as unknown as string).getTime();
        if (isNaN(eventTime) || eventTime < filters.since.getTime()) {
          return false;
        }
      }
      return true;
    });
  }

  public export_for_websocket(): Record<string, unknown>[] {
    // 假设 event 对象本身结构已经是可以直转 JSON 的类型
    return this._events.map(event => JSON.parse(JSON.stringify(event)));
  }

  public get count(): number {
    return this._events.length;
  }

  public clear(): void {
    this._events = [];
  }
}
